core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala

```diff
@@ -17,76 +17,175 @@
-    val providers = Seq(new HadoopFSDelegationTokenProvider(fileSystems)) ++
-      safeCreateProvider(new HiveDelegationTokenProvider) ++
-      safeCreateProvider(new HBaseDelegationTokenProvider)
+  /**
+   * Start the token renewer. Upon start, if a principal has been configured, the renewer will:
+   *
+   * - log in the configured principal, and set up a task to keep that user's ticket renewed
+   * - obtain delegation tokens from all available providers
+   * - schedule a periodic task to update the tokens when needed.
+   *
+   * When token renewal is not enabled, this method will not start any periodic tasks. Instead, it
+   * will generate tokens if the driver ref has been provided, update the current user, and send
+   * those tokens to the driver. No future tokens will be generated, so when that initial set
+   * expires, the app will stop working.
+   *
+   * @param driverEndpoint If provided, the driver where to send the newly generated tokens.
+   *                       The same ref will also receive future token updates unless overridden
+   *                       later.
+   * @return The newly logged in user, or null if a principal is not configured.
+   */
+  def start(driverEndpoint: Option[RpcEndpointRef] = None): UserGroupInformation = {
+    driverEndpoint.foreach(setDriverRef)
 
-    // Filter out providers for which spark.security.credentials.{service}.enabled is false.
-    providers
-      .filter { p => isServiceEnabled(p.serviceName) }
-      .map { p => (p.serviceName, p) }
-      .toMap
+    if (principal != null) {
+      renewalExecutor =
+        ThreadUtils.newDaemonSingleThreadScheduledExecutor("Credential Renewal Thread")
+
+      val originalCreds = UserGroupInformation.getCurrentUser().getCredentials()
+      val ugi = doLogin()
+
+      val tgtRenewalTask = new Runnable() {
+        override def run(): Unit = {
+          ugi.checkTGTAndReloginFromKeytab()
+        }
+      }
+      val tgtRenewalPeriod = sparkConf.get(KERBEROS_RELOGIN_PERIOD)
+      renewalExecutor.scheduleAtFixedRate(tgtRenewalTask, tgtRenewalPeriod, tgtRenewalPeriod,
+        TimeUnit.SECONDS)
+
+      val creds = obtainTokensAndScheduleRenewal(ugi)
+      ugi.addCredentials(creds)
+
+      val driver = driverRef.get()
+      if (driver != null) {
+        val tokens = SparkHadoopUtil.get.serialize(creds)
+        driver.send(UpdateDelegationTokens(tokens))
+      }
+
+      // Transfer the original user's tokens to the new user, since it may contain needed tokens
+      // (such as those user to connect to YARN). Explicitly avoid overwriting tokens that already
+      // exist in the current user's credentials, since those were freshly obtained above
+      // (see SPARK-23361).
+      val existing = ugi.getCredentials()
+      existing.mergeAll(originalCreds)
+      ugi.addCredentials(existing)
+      ugi
+    } else {
+      driverEndpoint.foreach { driver =>
```

Comment:
This is the part of the current API that I dislike the most, but I really don't know how to make it better without fixing SPARK-25689 first.

This at least keeps a single method to initialize delegation tokens, instead of having logic in the call sites about what method to call.

Reply:
ok so we agree that is a bit ugly and maybe there isn't a better api currently -- but are you opposed to the other suggestions, till the other changes happen?  I know its a small thing, I just think it would help keep the usage clear (or another set of asserts to resolve the ambiguity another way is fine too, I just don't like the ambiguity).